lib/pengine/unpack.c

```diff
@@ -4321,15 +4321,77 @@ mask_probe_failure(struct action_history *history, int orig_exit_status,
                       history->rsc->cluster);
 }
 
+/*!
+ * \internal Check whether a given failure is for a given pending action
+ *
+ * \param[in] history       Parsed history entry for pending action
+ * \param[in] last_failure  Resource's last_failure entry, if known
+ *
+ * \return true if \p last_failure is failure of pending action in \p history,
+ *         otherwise false
+ * \note Both \p history and \p last_failure must come from the same
+ *       lrm_resource block, as node and resource are assumed to be the same.
+ */
+static bool
+failure_is_newer(const struct action_history *history,
+                 const xmlNode *last_failure)
+{
+    guint failure_interval_ms = 0U;
+    long long failure_change = 0LL;
+    long long this_change = 0LL;
+
+    if (last_failure == NULL) {
+        return false; // Resource has no last_failure entry
+    }
+
+    if (!pcmk__str_eq(history->task,
+                      crm_element_value(last_failure, XML_LRM_ATTR_TASK),
+                      pcmk__str_none)) {
+        return false; // last_failure is for different action
+    }
+
+    if ((crm_element_value_ms(last_failure, XML_LRM_ATTR_INTERVAL_MS,
+                              &failure_interval_ms) != pcmk_ok)
+        || (history->interval_ms != failure_interval_ms)) {
+        return false; // last_failure is for action with different interval
+    }
+
+    if ((pcmk__scan_ll(crm_element_value(history->xml, XML_RSC_OP_LAST_CHANGE),
+                       &this_change, 0LL) != pcmk_rc_ok)
+        || (pcmk__scan_ll(crm_element_value(last_failure,
+                                            XML_RSC_OP_LAST_CHANGE),
+                       &failure_change, 0LL) != pcmk_rc_ok)
+        || (failure_change < this_change)) {
```

Comment:
Looks good assuming this is logically valid. I was surprised to learn that the history entry stays "pending" when it fails, so I'm not making any assumptions.

This seems reasonable though.

Reply:
Yeah I tried to clean up the state determination code a bit but it could definitely use more attention. If there is no recovery, both the pending and last_failure entries will be present, and the scheduler will know it's failed. If there is recovery, the stop will include scheduling a cleanup and cancelling recurring monitors, so both the pending and last_failure entries will be erased.